prime-router/src/test/kotlin/azure/BlobAccessTests.kt

```diff
@@ -52,4 +73,237 @@ class BlobAccessTests {
             assertThat(e).isNotNull()
         }
     }
+
+    @Test
+    fun `upload report`() {
+        val testUrl = "http://uploadreport"
+        val testFormat = Report.Format.CSV
+        val testBytes = "testbytes".toByteArray()
+
+        val one = Schema(name = "one", topic = Topic.TEST)
+        val metadata = Metadata(schema = one)
+        val report1 = Report(
+            one, listOf(listOf("1", "2"), listOf("3", "4")), source = TestSource,
+            bodyFormat = testFormat, metadata = metadata
+        )
+
+        mockkClass(BlobAccess::class)
+        mockkObject(BlobAccess.Companion)
+        every {
+            BlobAccess.uploadBody(
+                report1.bodyFormat, testBytes, report1.name, null,
+                Event.EventAction.NONE
+            )
+        } returns
+            BlobAccess.BlobInfo(report1.bodyFormat, testUrl, BlobAccess.sha256Digest(testBytes))
+
+        val testBlob = BlobAccess()
+        val result = testBlob.uploadReport(report1, testBytes)
+
+        assertThat(result.format).isEqualTo(testFormat)
+        assertThat(result.blobUrl).isEqualTo(testUrl)
+        assertThat(result.digest).isEqualTo(BlobAccess.sha256Digest(testBytes))
```

Comment:
Might be able to make this a little more concise by making a BlobInfo object in this test function, using that in the `every {} returns blobInfo`, and then comparing the full result object instead of comparing each field individually.

Reply:
For part of the test I wanted to simulate the actual data flow as much as possible, thus the mocked BlobInfo object being returned being constructed using the format from the input report instead of a prebuilt BlobInfo. I'm not really sure if anything is actually being tested if I did otherwise. I tried adjusting the comparison to a BlobInfo object built directly from the test variables, but apparently Kotlin tests equality by reference this way so the assertion fails; I could test the hashcodes of each instance of the data class instead, but testing the elements directly seemed more immediately readable.

Reply:
That works, but it doesn't work if I keep the original return value of the mock. From reading up on it I think it's because it's a data class where one of the elements of the class is an Array. All the same, if you're OK with the test in its current form I think it'd be best to just leave it as is.

Reply:
Yep, that's fine, and that makes sense about the Array, didn't think about that! Closing this convo.